From e0115e937c6df5672b05fd8464830e0ec19e6aa7 Mon Sep 17 00:00:00 2001
From: Matthew Lugg <mlugg@mlugg.co.uk>
Date: Mon, 10 Nov 2025 12:56:35 +0000
Subject: [PATCH 053/133] compiler: warn when using -fincremental with LLVM
 backend
X-Developer-Signature: v=1; a=openpgp-sha256; l=1147; i=xtex@astrafall.org;
 h=from:subject; bh=7jPobD00U19/1LmDNPFt55aECUZOm8oqml6lYoOh8fI=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78hjfWfZfjV8Tl7nt3l4RkZySCjvf+cKrPxjEz766/
 3pOgcrMjoksDGJcDJZiiixFhg3erDrp/KLLymVh5rAygQyRFmlgYGBgYGHgy03MKzXSMdIz1TbU
 MzTUMdIxYuDiFICp3lzKyHAwqy0sYVpL4J4e8+IJiW5/2Rf9Nb4kcN8gs8f0zVOn9wEM/9TWLb0
 k6H39Zpih04wMLZZPKu1fJj6X1txc5/Ljs+FrNRYA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

This configuration hasn't had much work put into it yet, so is all but
guaranteed to miscompile or crash. Since users are starting to try out
`-fincremental`, and LLVM is still the default backend in many cases,
it's worth having this warning to avoid bug reports like
https://github.com/ziglang/zig/issues/25873.

(cherry picked from commit 13993c4f62d8603aa5a25de70e2d111a01cbfaab)
---
 src/main.zig | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/main.zig b/src/main.zig
index 036adbf852b6..fd989cdc569d 100644
--- a/src/main.zig
+++ b/src/main.zig
@@ -3382,6 +3382,10 @@ fn buildOutputType(
         fatal("--debug-incremental requires -fincremental", .{});
     }
 
+    if (incremental and create_module.resolved_options.use_llvm) {
+        warn("-fincremental is currently unsupported by the LLVM backend; crashes or miscompilations are likely", .{});
+    }
+
     const cache_mode: Compilation.CacheMode = b: {
         // Once incremental compilation is the default, we'll want some smarter logic here,
         // considering things like the backend in use and whether there's a ZCU.
-- 
2.52.0

